### PR TITLE
APPSRE-3207: Support `use_channel_in_image_tag`

### PR DESCRIPTION
**Pull request by 2uasimojo**
Previously some consumers were using `IMAGE_TAG` as only *part* of the actual image tag referenced by a saas template, e.g.:

`  image: ${REGISTRY_IMG}:${CHANNEL}-${IMAGE_TAG}`

This was clearly wrong and misleading.

With this change, if the saas file/template specifies the parameter:

`  USE_CHANNEL_IN_IMAGE_TAG: "true"` (note: `"true"`, not `true`)

...then we will calculate the value of `IMAGE_TAG` to include the channel; so the above example can be changed to:

`  image: ${REGISTRY_IMG}:${IMAGE_TAG}`

This also makes it possible for `REPO_DIGEST` to be used by such consumers. They would specify:

```
parameters:
- name: REGISTRY_IMG
  required: true
- name: CHANNEL
  required: true
- name: USE_CHANNEL_IN_IMAGE_TAG
  value: "true"
- name: IMAGE_TAG
  required: true
- name: REPO_DIGEST
  required: true

...

        image: ${REPO_DIGEST}
```

APPSRE-3207

### CODE REVIEW

**Comment by 2uasimojo** [comment]
@jharrington22 @maorfr

**Comment by 2uasimojo** [comment]
Note to reviewers: Two commits here. Easier to review separately.

**Comment by jharrington22** [comment]
The additional `USE_CHANNEL_IN_IMAGE_TAG` parameter in the template allows consumers to be specific about whether or not channel is used in the digest fetch which keeps this change to just SREP or specific consumers (can't be added by mistake).

Long term when we switch to OPM builds for operators the catalog image channels go away. 

/lgtm

**Review by maorfr — CHANGES_REQUESTED** [review]
thanks again for this contribution!

this PR adds behavior that is based on parameters. it would be better to control behavior through the fields in the saas file itself rather then the parameters.

i'd look at `compare` as an example (search app-interface for `compare`).

**Comment by jharrington22** [comment]
> thanks again for this contribution!
> 
> this PR adds behavior that is based on parameters. it would be better to control behavior through the fields in the saas file itself rather then the parameters.
> 
> i'd look at `compare` as an example (search app-interface for `compare`).

@maorfr this is similar to `compare` no? Its switching on the digest request to use the channel parameter as part of the image name. Its not an ideal solution but given the time restrictions we don't see a cleaner way. Do you have a suggestion?

**Comment by 2uasimojo** [comment]
> it would be better to control behavior through the fields in the saas file

Done, I think. How does that look?

**Comment by jharrington22** [comment]
> > thanks again for this contribution!
> > this PR adds behavior that is based on parameters. it would be better to control behavior through the fields in the saas file itself rather then the parameters.
> > i'd look at `compare` as an example (search app-interface for `compare`).
> 
> @maorfr thanks for the feedback! This is similar to `compare` no? Its switching on the digest request to use the channel parameter as part of the image name. Its not an ideal solution but given the time restrictions we don't see a cleaner way. Do you have a suggestion?

I see move the control into the saas file. Sorry @maorfr I understand now don't mind me.

File: reconcile/utils/saasherder.py [hunk 1]
```diff
@@ -387,48 +406,58 @@ def _process_template(self, options):
...
+
+            # add IMAGE_TAG only if it is required and unspecified
+            if self._parameter_required_and_absent(
+                    "IMAGE_TAG", consolidated_parameters, template):
+                sha_substring = commit_sha[:hash_length]
+                # IMAGE_TAG takes one of two forms:
+                # - If saas file attribute 'use_channel_in_image_tag' is true,
+                #   it is {CHANNEL}-{SHA}
+                # - Otherwise it is just {SHA}
+                if self._get_saas_file_attribute("use_channel_in_image_tag"):
```
Comment (maorfr) [inline reconcile/utils/saasherder.py:419]: You'll need to add this field to the graphql query and set it in the saasherder init I believe?
It would also require a schema change in app-interface to support this change.

File: reconcile/utils/saasherder.py [hunk 2]
```diff
@@ -334,6 +334,25 @@ def _additional_resource_process(resources, html_url):
...
+                calculated.
+        :param template: The template file in dict form.
+        :return bool: True if the named parameter is required by the template,
+                but not already present in consolidated_parameters.
+        """
+        if parameter_name in consolidated_parameters:
+            return False
+        for template_parameter in template.get("parameters", {}):
+            if template_parameter["name"] == parameter_name:
+                return template_parameter.get("required", False)
```
Comment (maorfr) [inline reconcile/utils/saasherder.py:353]: This is new logic. We can't rely on the IMAGE_TAG being required, but we still want to generate it. Can we keep this to a separate PR?

Comment (2uasimojo) [reply reconcile/utils/saasherder.py:353]: Reverted the logic and left a TODO.

File: reconcile/utils/saasherder.py (hunk 1 above)
Comment (2uasimojo) [reply reconcile/utils/saasherder.py:419]: Found the query thing. Will look for the schema.

Comment (2uasimojo) [reply reconcile/utils/saasherder.py:425]: https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/17433 for the schema and documentation change. I suppose that should be merged with/after promoting this PR.

File: reconcile/utils/saasherder.py [hunk 3]
```diff
@@ -453,7 +488,7 @@ def _process_template(self, options):
                 logging.error(
                     f"[{url}/{path}:{target_ref}] " +
                     f"error fetching directory: {str(e)}")
-                return None, None
+                return None, None, None
```
Comment (maorfr) [inline reconcile/utils/saasherder.py:491]: Why is this needed?

Comment (2uasimojo) [reply reconcile/utils/saasherder.py:491]: Prevents these error conditions from raising `ValueError: not enough values to unpack (expected 3, got 2)` from the [caller](https://github.com/app-sre/qontract-reconcile/blob/035460ca61d130aab188db6359c602db1b2bf92d/reconcile/utils/saasherder.py#L682-L683).

Comment (2uasimojo) [reply reconcile/utils/saasherder.py:491]: (That's the "latent buglet" referenced by the [first commit](https://github.com/app-sre/qontract-reconcile/pull/1486/commits/aa54ee96618804f0f2991e044efd973719154348))

Comment (maorfr) [reply reconcile/utils/saasherder.py:491]: OH NICE

File: reconcile/utils/saasherder.py [hunk 4]
```diff
@@ -334,6 +334,31 @@ def _additional_resource_process(resources, html_url):
...
                         resource['name'])
 
+    @staticmethod
+    def _parameter_required_and_absent(
+            parameter_name, consolidated_parameters, template):
+        """Is a parameter is required in the template but unspecified?
+
+        TODO: This is currently "parameter *named* and absent" -- i.e. we
+        don't care about `required: true`. This is for backward compatibility;
+        but should eventually be fixed.
```
Comment (maorfr) [inline reconcile/utils/saasherder.py:344]: i'm not sure this statement is true. i think it's enough to name the parameter. most services just use the value `latest` and not make it required, which helps them deploy the template in their development process.

File: reconcile/utils/saasherder.py (hunk 1 above)
Comment (maorfr) [reply reconcile/utils/saasherder.py:425]: :heart:

**Review by maorfr — APPROVED** [review]
otherwise for that comment, LGTM

this is very nice work. very involved and you are doing it effortlessly and with very little hints.
kudos!

File: reconcile/utils/saasherder.py (hunk 4 above)
Comment (2uasimojo) [reply reconcile/utils/saasherder.py]: I renamed the method and took out the TODOs.

(Nitpick: If they specify the value, `required` is redundant/unnecessary and wouldn't trigger this function anyway. But after poking around, I think I grok why it will *never* be possible to track down all the template files in existence and make sure they've specified `required: true` for `IMAGE_TAG`.)

Comment (maorfr) [reply reconcile/utils/saasherder.py]: EXACTLY! if we can't automate it, we can't enforce it.
